feat: re-write solid bindings with multistore support

bigmistqke opened the pull request:
continuation of https://github.com/livestorejs/livestore/pull/783

## Table of Contents

- [API Overview](#api-overview)
  - [Setup](#setup)
  - [`useStore`](#usestore)
    - [`store.useQuery`](#storeusequery)
    - [`store.useClientDocument`](#storeuseclientdocument)
    - [Suspense Behavior](#suspense-behavior)
  - [`useQuery`](#usequery)
  - [`useClientDocument`](#useclientdocument)
  - [`useStoreRegistry`](#usestoreregistry)
  - [`withSolidApi`](#withsolidapi)
  - [`LiveList`](#livelist)
- [Testing Setup](#testing-setup)
- [Changes in Other Packages](#changes-in-other-packages)
  - [`mono test` CLI](#mono-test-cli)
  - [`@livestore/framework-toolkit`](#livestoreframework-toolkit-new)
  - [`@livestore/common`](#livestorecommon)
- [Demos](#demos)

## API Overview

### Setup

```tsx
import { StoreRegistry } from '@livestore/livestore'
import { StoreRegistryProvider } from '@livestore/solid'
import { batch } from 'solid-js'

const storeRegistry = new StoreRegistry({ defaultOptions: { batchUpdates: batch } })

const App = () => (
  <StoreRegistryProvider storeRegistry={storeRegistry}>
    <MyComponent />
  </StoreRegistryProvider>
)
```

### `useStore`

Returns 
- a Resource that suspends on read.
- SolidApi:
    - [store.useQuery](#storeusequery)
    - [store.useClientDocument](#storeuseclientdocument)

```tsx
function Issue(props: { issueId: string }) {
  const store = useStore(() => issueStoreOptions(props.issueId))

return (
    <Show when={store()}>{(s) => <IssueView store={s()} />}</Show>
  )
}
```

When `useStore()` options are reactive (e.g., accessor), `store.useQuery` and `store.useClientDocument` automatically switch to the new store:

```tsx
function Issue(props: { issueId: string }) {
  // When props.issueId changes, store switches to new instance
  const store = useStore(() => issueStoreOptions(props.issueId))

  // Queries automatically re-subscribe to the new store
  const comments = store.useQuery(queryDb(tables.comments.select()))
  const [draft, setDraft] = store.useClientDocument(tables.draft)

  return <CommentList comments={comments()} />
}
```

#### `store.useQuery`

Subscribes to query. Accepts `Accessor<T> | T`.

```tsx
const todos = store.useQuery(queryDb(tables.todos.select()))
const filtered = store.useQuery(() => queryDb(tables.todos.where({ status: filter() })))
```

#### `store.useClientDocument`

Signal-like API. Returns `[state, setState, id, query$]`. To enable destructuring at declaration, state can be set before store is initialized—updates are buffered locally and applied once ready.

```tsx
const [state, setState] = store.useClientDocument(UIState)
setState({ sidebarOpen: !state().sidebarOpen })
```

#### Suspense Behavior

Both `store.useQuery` and `store.useClientDocument` eagerly evaluate store internally (with a memo). This triggers Suspense at their declaration, not from where they are read. This might create suspense behaviors that feel a bit counter-intuitive. 

I experimented with custom lazy memoization in an earlier commit, but reverted it as it caused cleanup bugs in the tests.

##### Example 1: useStore

useStore suspends at read: `child` will be initially displayed

```tsx
const Child = () => {
  const store = useStore()
  return <Suspense fallback="child">{store()?.id}</Suspense>
}
const App = () => <Suspense fallback="root"><Child></Suspense>
```

##### Example 2: useStore + store.useQuery

store.useQuery eagerly evaluates the store: `root` will be initially displayed

```tsx
const Child = () => {
  const store = useStore()
  const query = store.useQuery(...)
  return <Suspense fallback="child">{query()?.id}</Suspense>
}
const App = () => <Suspense fallback="root"><Child></Suspense>
```

##### Example 3: useStore + store.useClientDocument

store.useClientDocument eagerly evaluates the store: `root` will be initially displayed

```tsx
const Child = () => {
  const store = useStore()
  const [state] = store.useClientDocument(...)
  return <Suspense fallback="child">{state()}</Suspense>
}
const App = () => <Suspense fallback="root"><Child></Suspense>
```

### `useQuery`

Standalone hook when you have a store instance.

```tsx
const todos = useQuery(
  queryDb(tables.todos.select()),
  { get store() { return props.store } }  // getter for reactivity
)
```

### `useClientDocument`

Standalone hook when you have a store instance.

```tsx
const [state, setState] = useClientDocument(
  UIState, 'singleton', undefined,
  { get store() { return props.store } }  // getter for reactivity
)
```

### `useStoreRegistry`

Access registry for preloading.

```tsx
const storeRegistry = useStoreRegistry()
storeRegistry.preload(issueStoreOptions(props.issueId))
```

### `withSolidApi`

Augments a Store with Solid hooks. Called automatically by `useStore()`.

```tsx
const storeWithApi = withSolidApi(existingStore)
```

### `LiveList`

Optimized list rendering.

```tsx
<LiveList store={store()} query={query} renderItem={(todo) => <TodoItem todo={todo} />} />
```

## Testing Setup

Solid uses resolve conditions for reactive (client) vs non-reactive (SSR) primitives. Unlike React, we can't use Node for both—requires separate vitest configs and fetch mocking (like Svelte).

```
├── vitest.config.client.ts   # jsdom + ['browser', 'solid']
├── vitest.config.server.ts   # node + ['node', 'solid']
├── *.client.test.tsx         # reactive tests
└── *.server.test.tsx         # SSR tests
```

## Changes in Other Packages

### `mono test` CLI

Updated to support Solid's separate client/SSR test configs:
- Discovers multiple `vitest*.config.ts` per package, runs each with `--config`
- Added `root: import.meta.dirname` to configs for workspace-root discovery

Note: I tried using a single `vitest.config.ts` with `test: { projects: [...] }` but this caused bugs—resolve conditions were not respected and client code incorrectly detected a Node environment.

### `@livestore/framework-toolkit` (new)

Shared utilities extracted for React and Solid adapters:
- **Query**: `createQueryResource`, `runInitialQuery`, `computeRcRefKey`, `normalizeQueryable`, `formatQueryError`, `getResourceLabel`
- **Client document**: `validateTableOptions`, `removeUndefinedValues`
- **Stack info**: `captureStackInfo`, `originalStackLimit`, `StackInfo`
- **Types**: `Dispatch`, `SetStateAction`, `SetStateActionPartial`, `StateSetters`, `NormalizedQueryable`

### `@livestore/common`
- Added `exposeStoreForDebugging`

## Demos

- `examples/web-todomvc-solid` — TodoMVC with multi-store setup
- `examples/cf-chat-solid` — Real-time chat with Cloudflare sync

bigmistqke committed:
feat: initial port of react api to solid

bigmistqke committed:
feat: simplify signature setState

bigmistqke committed:
feat: port tests from react to solid

thx claude

bigmistqke committed:
feat: add support for queryDef to be Accessor<T> | T instead of Accessor<T> only

removes the need for unnecessary wrapping in callbacks + in line with the signature of solid 2.0's createSignal.

bigmistqke committed:
feat: make table- and id- argument Accessor<T> | T instead of Accessor<T>

bigmistqke committed:
chore: cleanup LiveStoreProvider

bigmistqke committed:
fix: fix cleanup in useQueryRef

bigmistqke committed:
fix: fix reactivity for queryRefOptions

bigmistqke committed:
chore: add @solid-primitives/deep

bigmistqke committed:
test: update snapshots

bigmistqke committed:
chore: remove solid references in react's tsconfig

bigmistqke committed:
config: add adapter-web and utils-dev project references to tsconfig

bigmistqke committed:
refactor: remove useRcResource and update reactive hooks

- Remove deprecated useRcResource hook and its tests
- Remove unused create-writable utility
- Update useQuery and useClientDocument implementations
- Update related test files

bigmistqke committed:
refactor: convert LiveList component from React to Solid.js

- Replace React hooks with Solid.js primitives
- Update type imports and component structure
- Simplify component implementation for Solid.js patterns

bigmistqke committed:
chore: update imports and test configurations

- Update import paths to use .js extensions
- Fix role attributes in test components
- Update vitest configuration imports
- Minor code cleanup and formatting

bigmistqke committed:
test: update snapshots for useQuery and useClientDocument tests

- Update test snapshots after hook refactoring
- Reflect changes from useRcResource removal

bigmistqke committed:
test: fix role attributes in test components

- Change role="content" to role="group" in test components
- Change role="current-id" to role="group" to use valid ARIA roles

bigmistqke committed:
test: use type casting for non-standard role attributes in tests

- Cast 'content' and 'current-id' roles as any to bypass TypeScript validation
- Use Show component instead of ternary for conditional rendering

bigmistqke committed:
test: fix should work for a dynamic list with query swapping

bigmistqke committed:
chore: update pnpm-lock.yaml

bigmistqke committed:
feat: make useQuery accept Queryable

- Update useQuery to accept any Queryable type instead of just LiveQueryDef
- Add support for query builders, live queries, and signal definitions
- Normalize different queryable types to a common internal format
- Maintain backwards compatibility with existing LiveQueryDef usage

bigmistqke committed:
update snapshot useQuery

bigmistqke committed:
chore: update pnpm-lock.yaml

bigmistqke committed:
chore: update pnpm-workspace.yaml

bigmistqke committed:
chore: set solid dependency to catalog:

bigmistqke committed:
fix: resolve merge conflict markers in package.json files

bigmistqke committed:
config: disable React hook rules for Solid.js package in biome

bigmistqke committed:
refactor(solid): use StoreInternalsSymbol for accessing store internals

- Update useQuery.ts and useClientDocument.ts to use StoreInternalsSymbol
- Add useLiveStoreContext hook for accessing store within LiveStoreProvider
- Add StoreRegistryContext and StoreRegistryProvider for multi-store support
- Remove deprecated store.ts module
- Update tests to use StoreInternalsSymbol for reactivityGraph access
- Update snapshots

bigmistqke committed:
refactor(solid): replace LiveStoreProvider with StoreRegistry pattern

- Remove LiveStoreProvider.tsx (replaced by StoreRegistryContext.tsx)
- Remove LiveStoreProvider.test.tsx (redundant with new pattern)
- Add useStore.test.tsx ported from React for StoreRegistry tests
- Tests adapted for Solid's reactive patterns (signals vs rerender)

bigmistqke committed:
feat(solid): add SSR and client test separation

Following solid-lib-starter pattern:
- SSR tests use renderToString in node environment (test/server.test.tsx)
- Client tests use jsdom globals in node environment for WASM compatibility
- Separate test modes: `vitest run` for client, `vitest run --mode ssr` for SSR

bigmistqke committed:
test(solid): implement skipped Suspense boundary tests

Enable previously skipped useStore tests for Suspense integration:
- works with Suspense boundary
- does not re-suspend on subsequent renders when store is already loaded

The unusedCacheTime=0 test remains skipped (see issue #916).

bigmistqke committed:
chore(solid): remove unused LiveStoreContext

bigmistqke committed:
feat(solid): export LiveList from main module

bigmistqke committed:
style(solid): format useStore.test.tsx

bigmistqke committed:
refactor(solid): use namespace import for solid-js

bigmistqke committed:
refactor(solid): use createComputed for side effects instead of createMemo/createRenderEffect

bigmistqke committed:
refactor: extract shared TodoMVC fixture to tests/package-common

Move common test schema, tables, events, and store creation logic
to a shared package that can be used by both React and Solid tests.
This reduces duplication and ensures consistency across framework tests.

bigmistqke committed:
refactor: extract shared framework utilities into @livestore/framework-toolkit

- Create new @livestore/framework-toolkit package for shared code between React/Solid
- Move query utilities (normalizeQueryable, computeRcRefKey, formatQueryError)
- Move client-document utilities (validateTableOptions, removeUndefinedValues)
- Move stack-info utilities (captureStackInfo, originalStackLimit, StackInfo type)
- Move todomvc-fixture to framework-toolkit/testing
- Update React and Solid packages to use framework-toolkit
- Re-export types for backwards compatibility

bigmistqke committed:
refactor(framework-toolkit): add createQueryResource to reduce boilerplate

Extract span creation and queryRcRef initialization into shared createQueryResource function.
Also adds getResourceLabel and runInitialQuery helpers.

Updates React and Solid useQuery to use the new shared utilities.

bigmistqke committed:
test(solid): split tests into client and server files

Restructure tests to separate client-side (DOM) tests from server-side (SSR) tests:
- *.client.test.tsx: Use render() from @solidjs/testing-library
- *.server.test.tsx: Use renderToString() from solid-js/web

Update vitest.config.ts to run:
- Client mode (default): only *.client.test.tsx files
- SSR mode (--mode ssr): only *.server.test.tsx files

This fixes the 'Client-only API called on server side' errors by ensuring
proper test isolation between client and server environments.

bigmistqke committed:
fix(solid): fix vitest workspace mode resolving SSR builds instead of browser

Split vitest config into separate client and SSR configs to properly handle
Solid's different JSX transforms and module resolution requirements.

Key changes:
- Add vitest.client.config.ts with browser conditions and server.deps.inline
  for solid-js to ensure browser builds are used in jsdom environment
- Add vitest.ssr.config.ts with node conditions and a plugin to remove
  browser condition injected by vite-plugin-solid's configEnvironment hook
- Update vite-plugin-solid to 2.11.10 which respects ssr option in Vite 6+
- Add WASM fetch shim in setup.ts for wa-sqlite compatibility in jsdom
- Fix SSR test assertion to account for Solid's hydration markers

The core issue was that when running tests from the workspace root,
vite-plugin-solid's configEnvironment hook (Vite 6+) was adding 'browser'
to resolve conditions even for SSR tests, causing solid-js to resolve
to server builds which throw "Client-only API called on server side".

bigmistqke committed:
docs(solid): fix incorrect API usage in solid integration examples

- Change `getStore` to `useStore` (correct Solid API)
- Change `query()` to `store().useQuery()` (correct pattern)
- Create `useAppStore()` wrapper hook matching React pattern

bigmistqke committed:
chore(biome): disable React hook rules for Solid documentation snippets

Add biome.jsonc override to disable useHookAtTopLevel and
useExhaustiveDependencies for Solid docs directories. This removes
the need for inline biome-ignore comments since Solid's reactive
patterns don't follow React's hook rules.

bigmistqke committed:
style(solid): apply linter formatting to vitest configs

bigmistqke committed:
chore(deps): upgrade vite-plugin-solid to 2.11.10

Fixes SSR condition handling in Vite 6+ where the configEnvironment
hook was incorrectly adding 'browser' condition.

bigmistqke committed:
refactor(common): add exposeStoreForDebugging helper

Extract duplicate debug store exposure logic into a shared helper
in @livestore/common. Update React and Solid useStore hooks to use it.

bigmistqke committed:
refactor(livestore): use exposeStoreForDebugging helper in LiveStore.ts

Replace 3 duplicate inline debug store exposure blocks with the
shared helper from @livestore/common.

bigmistqke committed:
refactor(solid): simplify useStore by moving retain into resource callback

Move storeRegistry.retain() call into the createResource callback,
removing the need for separate createMemo and createComputed.

bigmistqke committed:
refactor(solid): fix useClientDocument signature and type inference

- Remove misleading `idOrOptions` parameter name (vestige from old useRow API)
- Add UseClientDocument interface that enforces id requirement based on table definition:
  - Tables with default id: id parameter is optional
  - Tables without default id: id parameter is required
- Make store a required parameter instead of optional with runtime assertion
- Rename `idStr` to `serializedId` for clarity
- Remove unused `when` import from solid-whenever
- Simplify withSolidApi using Object.assign with satisfies

bigmistqke committed:
refactor(test): support multiple vitest configs per package

- Add `root: import.meta.dirname` to Solid vitest configs so tests are
  found when run from workspace root
- Update test discovery to find all vitest config files in packages
- Run each config separately with `--config` flag instead of passing
  multiple paths to single vitest invocation
- Support `--filter` option discovering and running multiple configs

bigmistqke committed:
refactor(react): add root to vitest config for consistent test discovery

bigmistqke committed:
test(react): update useClientDocument snapshot for stack info changes

bigmistqke committed:
refactor(vitest): add root to remaining vitest configs

Add `root: import.meta.dirname` to docs, common-cf, and sqlite-wasm
vitest configs for consistent test discovery from workspace root.

bigmistqke committed:
refactor(solid): add explicit store prop to LiveList component

Pass store configuration to useQuery calls instead of relying on
context, aligning with the updated useQuery signature.

bigmistqke committed:
chore(solid): remove @bigmistqke/solid-whenever dependency

bigmistqke committed:
fix(solid): conditionally pass default option to table.get

bigmistqke committed:
refactor(solid): update todomvc example to use new useStore API

- Use StoreRegistryProvider, Suspense, ErrorBoundary in index.tsx
- Update store.ts to export useAppStore hook
- Update components to use store().useQuery() pattern
- Rename ActionBar to Footer for consistency with React example
- Add biome override to disable React hook rules for Solid example

bigmistqke committed:
refactor(solid): add local state buffering to useStore.useClientDocument

bigmistqke committed:
refactor(solid): update todomvc components to use new useStore API

bigmistqke committed:
feat(examples): add cf-chat-solid example

bigmistqke committed:
docs(solid): update MainSection snippet to use new useStore API

bigmistqke committed:
test(solid): add tests for useStore.useQuery and useStore.useClientDocument

- Add useQuery tests: supports query builders directly, union of different result types
- Add useStore.useQuery tests: returns undefined before store loads, updates reactively
- Add useStore.useClientDocument tests: state accessor/setter, multiple updates, buffering, raw commits
- Fix buffering bug: clear localState after syncing to prevent re-sync on memo re-runs

bigmistqke committed:
test(solid): add tests for useQuery and useClientDocument store switching

bigmistqke committed:
refactor(solid): remove SolidResourceApi export from mod.ts

bigmistqke committed:
refactor(solid): support both Store and Accessor in withSolidApi

bigmistqke committed:
fix(solid): use optional chaining for useQuery results in tests

bigmistqke committed:
test(solid): update useClientDocument snapshot

bigmistqke committed:
refactor(solid): simplify useQuery in withSolidApi using createMemo

Changes Suspense behavior due to eager evaluation of the memo.

bigmistqke committed:
fix(solid): avoid Suspense when store is already cached

Return synchronously when getOrLoadPromise returns a Store instead of a Promise.

bigmistqke committed:
test(solid): add createSuspenseCount utility and refactor Suspense tests

- Add createSuspenseCount helper for tracking Suspense fallback renders
- Refactor useStore tests to verify Suspense triggers at correct boundaries
- Add explicit suspenseCount assertions to useQuery and useClientDocument tests

bigmistqke commented:
I played a bit with lazily suspending `store.useClientDocument` and `store.useQuery` in a separate branch: [solid-suspend-at-read](https://github.com/bigmistqke/livestore/commit/3e2d0f2a36ca9cd4e4936ce5f0776d5bb0256498). 

It's fighting the solid 1.X primitives a little bit, and requires a bit of hackery to make it work (especially the [bypassSuspense-utility](https://github.com/bigmistqke/livestore/blob/3e2d0f2a36ca9cd4e4936ce5f0776d5bb0256498/packages/%40livestore/solid/src/utils.tsx#L36) is uuugly), so i kept it in a separate branch for now.

## Differences in suspense behavior

```tsx
function Child(){
 const store = useStore(...);
 const query = store.useQuery(...);

 return <Suspense fallback="inner">{query()}</Suspense>
}

function App(){
 return <Suspense fallback="outer"><Child/></Suspense>
}
```

**differences**
- `solid-2` branch: `outer` will be displayed initially
- `solid-suspend-at-ready` branch: `inner` will be displayed initially

Maybe it's nice to have a voice chat at some point to figure out what would be the best/consistent pattern of suspense between the different frontend adapters.

schickling reviewed (commented):
## Code Review: Solid Bindings with Multistore Support

Great work on this comprehensive rewrite! The implementation follows idiomatic Solid patterns and the new `@livestore/framework-toolkit` package nicely reduces duplication between React and Solid adapters.

### Strengths

1. **Clean API Design**: The `useStore` hook returning a Resource with `.useQuery()` and `.useClientDocument()` methods is ergonomic for Solid developers.

2. **Good Code Sharing**: The `@livestore/framework-toolkit` extracts shared logic (query utilities, client document utilities, stack info) ensuring consistent behavior across adapters.

3. **Proper Solid Patterns**: Good use of `createResource`, `createMemo`, `onCleanup`, and reactive primitives.

4. **Smart State Buffering**: The `useClientDocument` design allowing `setState` before store loads is clever and enables flexible component patterns.

5. **Comprehensive Tests**: Good coverage of Suspense behavior, store switching, state buffering, and reactivity.

### Action Items Before Merge

1. **Inline `@bigmistqke/solid-whenever`**: We should inline the needed utilities (`when`, `every`) to avoid the extra dependency on a personal package. This keeps the dependency footprint minimal.

2. **Suspense Behavior Alignment**: @IGassmann - would love your feedback on the Suspense behavior here. Currently `store.useQuery` and `store.useClientDocument` eagerly evaluate the store (triggering Suspense at declaration), while `useStore()` alone suspends at read. The PR description documents this well, and there's an experimental branch exploring lazy suspension. What's your take on the best pattern for consistency?

3. **Minor Type Safety**: Consider using generics instead of `any` in `withSolidApi`'s `useClientDocument` method (`useStore.ts:150`).

4. **Skipped Test**: The `unusedCacheTime=0` test references issue #916 - should be addressed or documented as a known limitation.

5. **Example Missing `batchUpdates`**: `examples/web-todomvc-solid/src/index.tsx` doesn't pass `batchUpdates: batch` to `StoreRegistry` as shown in the setup docs.

### Next Steps

We'll rebase this PR to the latest `dev` branch shortly and get things ready for merge. The vite version mismatch errors in the build are unrelated to this PR and will be resolved during the rebase.

Overall this is high-quality work - just need to address the dependency inlining and align on Suspense behavior, then we're good to go! 🚀

schickling committed:
fix: align __debugLiveStore type declarations and remove unused @solid-primitives/deep dep

schickling committed:
refactor(solid): inline @bigmistqke/solid-whenever utilities to remove external dependency

Inlined the 'when' and 'every' utilities from @bigmistqke/solid-whenever
into a local whenever.ts file. This removes one external dependency from
the package, reducing bundle size and dependency footprint.

schickling committed:
fix: add missing vite-plugin-vue-devtools dep and update knip config for test files

- Added vite-plugin-vue-devtools to docs code snippets package to fix TS error
- Added .test.tsx pattern to knip ignore list
- Added test dependencies to knip ignoreDependencies

schickling committed:
Merge branch 'dev' into solid-2

schickling committed:
test(cf-chat-solid): add Playwright tests copied from cf-chat React example

schickling committed:
fix(vitest): add root to vitest configs to ensure tests are found when run from workspace root

schickling committed:
fix(solid): allow useClientDocument without explicit id parameter

The withSolidApi.useClientDocument was requiring id to be truthy via
every(store, table, id), but id can legitimately be undefined when the
table has a default id. This caused useClientDocument to return undefined
when no id was passed, breaking the cf-chat-solid example.

Also fixed ChatBot name in cf-chat-solid to include emoji to match tests.

IGassmann reviewed (commented):
Note: The top-level `useQuery` and `useClientDocument` exports were removed from `@livestore/react` in #947.

The rationale was to remove ambiguity about which store is being queried: standalone hooks require passing a store option, whereas `store.useQuery()` makes the association explicit at the call site. This becomes especially important in multi-store scenarios.

Should we follow the same approach here, or is there a Solid-specific reason to keep the standalone exports?

IGassmann reviewed (commented):
794a925b9 centralized `__debugLiveStore` registration in `create-store.ts` to avoid duplicate registration across framework adapters and added cleanup via `Effect.addFinalizer` to fix a memory leak where disposed stores remained in `__debugLiveStore`. Since registration now happens automatically in core, `exposeStoreForDebugging()` shouldn't be needed.

IGassmann commented on packages/@livestore/livestore/src/store/create-store.ts:
```diff
@@ -45,7 +45,7 @@ import { StoreInternalsSymbol } from './store-types.ts'
   /** Store instances for console debugging */
-  var __debugLiveStore: Record<string, Store<any, any>> | undefined
+  var __debugLiveStore: Record<string, unknown> | undefined
 }
```
Is this change from `Store<any, any>` to `unknown` intentional? #976 added the stronger Store typing.

IGassmann commented:
**Re: Suspense behavior alignment**

I don't have a strong opinion on the suspense behavior here as I'm not familiar enough with Suspense best practices in Solid to judge what feels most natural to Solid developers.

IMO, what matters most is that the solution is **idiomatic and optimized for the Solid context**. If eager evaluation via is the pattern that Solid developers expect and that works best with Solid's reactivity model, then that's the right choice.

Cross-framework consistency isn't particularly important here. It's rare for the same user to be using LiveStore with multiple frontend frameworks simultaneously. My opinion is that each adapter should prioritize what feels natural and performs well within its own ecosystem rather than artificially constraining itself to match other frameworks.

If the `solid-suspend-at-read` branch approach is more aligned with Solid conventions and the cleanup bugs can be resolved, that could be worth pursuing. But if the current eager evaluation approach is considered acceptable in the Solid community, I'd defer to that judgment.

schickling committed:
fix(solid): address Igor's feedback

1. Restore __debugLiveStore type to Store<any,any> (was changed to unknown)
2. Remove exposeStoreForDebugging from Solid/React adapters - registration
   is now centralized in create-store.ts (per 794a925b9)
3. Remove standalone useQuery/useClientDocument exports from Solid for API
   consistency with React (#947) - use store.useQuery() instead
4. Remove debug-store.ts from @livestore/common (was added by this branch
   but no longer needed since registration is centralized)
5. Restore LiveStore.ts from dev branch (had stale exposeStoreForDebugging)

schickling committed:
test(solid): update Suspense test assertions for suspend-at-read pattern

- Update useQuery test to expect Suspense at innermost boundary (UseQuerySuspense)
- Update useClientDocument test to expect no Suspense with optimistic updates
- Remove unused MakeOptional type from utils.tsx

schickling committed:
fix(solid): revert suspend-at-read pattern to fix reactivity issues

The bypassSuspense utility was breaking reactive updates when data synced
between clients. Reverted to the original implementation which properly
maintains Solid's reactive tracking.

- Revert useQuery and useClientDocument to original implementations
- Revert test assertions to match actual Suspense behavior
- Remove unused bypassSuspense utility
- Restore utils.ts (was utils.tsx)

schickling commented:
## Update: Suspend-at-Read Pattern Deferred

I attempted to implement the "suspend-at-read" pattern (where Suspense triggers when the accessor is read in JSX, not when `store.useQuery()` is called) to make the Solid adapter more idiomatic.

### What was tried

Used a `bypassSuspense` utility that wrapped Resource reads in an internal Suspense boundary via `Solid.children()` to defer suspension to the read site.

### What happened

The implementation broke reactive updates in multi-client sync scenarios. The `cf-chat-solid` Playwright test "user sidebar shows current user and others correctly" consistently failed - when Bob joined a chat room, Alice never saw Bob appear in the sidebar because the `users` query wasn't updating reactively.

### Resolution

Reverted to the original implementation which properly maintains Solid's reactive tracking. The current Suspense behavior (triggers at declaration) is documented in the PR description.

### Follow-up

Created #987 to track the suspend-at-read implementation as follow-up work. The issue contains:
- Detailed research on the pattern and why it matters
- The attempted implementation approach
- Root cause analysis of the reactivity bug
- Open questions for future investigation
- Acceptance criteria

The suspend-at-read pattern would be a nice-to-have for idiomatic Solid.js usage, but the current implementation is fully functional and well-tested.

schickling committed:
chore: cleanup and version bump

- Bump @livestore/framework-toolkit version to 0.4.0-dev.22
- Remove duplicate resolve function from whenever.ts (now imports from utils.ts)
- Unify on AccessorMaybe type from utils.ts

schickling committed:
fix(livestore): fix flaky StoreRegistry abort test

The test incorrectly assumed getOrLoadPromise always returns a Promise
after abort. In reality, it can return either a Store or Promise<Store>
depending on timing. The fix handles both cases correctly.

schickling commented:
Thanks a lot again for your nice contribution @bigmistqke! 🙏
